### PR TITLE
fix: wheel scroll error!

### LINKED ISSUE #2627: 透视表列多时，使用触摸板左右快速滑动，会报栈溢出错误
### 🏷 Version

| Package        | Version |
| -------------- | ------- |
| @antv/s2       |   next      |
| @antv/s2-react |         |
| @antv/s2-vue   |         |

### Sheet Type

- [X] PivotSheet
- [ ] TableSheet
- [ ] GridAnalysisSheet
- [ ] StrategySheet
- [ ] EditableSheet

### 🖋 Description

透视表列比较多，使用触摸板左右滑动快时会出现栈溢出错误, 看着像是@antv/g引擎报错了。
```
Uncaught RangeError: Maximum call stack size exceeded
    at @antv_s2.js?v=75bfa16a:39272:55
    at @antv_s2.js?v=75bfa16a:10425:18
    at Array.forEach (<anonymous>)
    at SyncHook2.call (@antv_s2.js?v=75bfa16a:10424:22)
    at RenderingService2.render (@antv_s2.js?v=75bfa16a:12154:29)
    at Canvas2.render (@antv_s2.js?v=75bfa16a:16213:24)
    at RootInteraction2.draw (@antv_s2.js?v=75bfa16a:43746:34)
    at RootInteraction2.clearState (@antv_s2.js?v=75bfa16a:43750:14)
    at RootInteraction2.changeState (@antv_s2.js?v=75bfa16a:43769:12)
    at PivotSheet2.<anonymous> (@antv_s2.js?v=75bfa16a:41177:21)
```

### ⌨️ Code Snapshots
官网demo向右滑动快的时候就能复现这个问题
https://s2.antv.antgroup.com/examples/case/performance-compare/#pivot

### 🔗 Reproduce Link

image

### 💻 System information

| Environment | Info |
| ------- | ------- |
| System  |         |
| Browser |         |

### PR DESCRIPTION
**Pull request by DavisYao88**
### 👀 PR includes

🐛 Bugfix

- [x] Solve the issue and close #2627

🔧 Chore

- [ ] Test case
- [x] Docs / demos update
- [ ] CI / workflow
- [ ] Release version
- [ ] Other ()

### 📝 Description

修复使用触摸板快速反复滚动时，堆内存溢出，导致表格卡死。
### 🖼️ Screenshot

| Before | After |
| ------ | ----- |
| f7a572ed5556fd4ca6c89fc7e9981271      | 8E764474-A98E-4ae2-BC14-C589818CB69C   |

### 🔗 Related issue link

close #2627

### 🔍 Self-Check before the merge

- [ ] Add or update relevant docs.
- [x] Add or update relevant demos.
- [ ] Add or update test case.
- [ ] Add or update relevant TypeScript definitions.

### CODE REVIEW

**Comment by EliazTray** [comment]
Any progress？

**Comment by DavisYao88** [comment]
> It looks like two months have passed, any progress here? Thanks~



> It looks like two months have passed, any progress here? Thanks~

**Comment by lijinke666** [comment]
不好意思，下周处理该 pr ，感谢贡献

**Comment by lijinke666** [comment]
@DavisYao88 请解决下冲突

File: packages/s2-react/playground/index.tsx [hunk 1]
```diff
@@ -1642,6 +1644,11 @@ function MainLayout() {
                 label: '绘制 G2 图表',
                 children: <ChartSheet />,
               },
+              {
+                key: 'bigData',
+                label: '100w数据',
```
Comment (lijinke666) [inline packages/s2-react/playground/index.tsx:1658]:
```suggestion
                label: '100w 数据',
```

File: packages/s2-react/playground/components/BigDataSheet.tsx [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+const s2DataConfig = {
+  fields: {
+    rows: ['type', 'subType'],
+    columns: ['province', 'city'],
+    values: ['number'],
+  },
+  data: generateRawData(
+    { province: 10, city: 100 },
+    { type: 10, sub_type: 100 },
```
Comment (lijinke666) [inline packages/s2-react/playground/components/BigDataSheet.tsx:57]: image

File: packages/s2-react/playground/components/BigDataSheet.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+          });
+        }
+      }
+    }
+  }
+
+  return res;
+}
+
+const s2DataConfig = {
```
Comment (lijinke666) [inline packages/s2-react/playground/components/BigDataSheet.tsx:49]:
```suggestion
const s2DataConfig: S2DataConfig = {
```

File: packages/s2-core/src/facet/base-facet.ts [hunk 4]
```diff
@@ -1249,6 +1254,40 @@ export abstract class BaseFacet {
   onWheel = (event: WheelEvent) => {
     const { interaction } = this.spreadsheet.options;
     let { deltaX, deltaY, offsetX, offsetY } = event;
+    const {
```
Comment (lijinke666) [inline packages/s2-core/src/facet/base-facet.ts]: 应该写在这之后, 不然父容器也会滚动, 需要验证滚动链的问题

https://github.com/antvis/S2/blob/b3d55708bea82142a81dd1d2ede6ff300d5e8957/packages/s2-core/src/facet/base-facet.ts#L1359-L1370

File: packages/s2-core/src/facet/base-facet.ts [hunk 5]
```diff
@@ -1249,6 +1254,40 @@ export abstract class BaseFacet {
...
+      scrollX: currentScrollX,
+      scrollY: currentScrollY,
+      rowHeaderScrollX,
+    } = this.getScrollOffset();
+
+    if (this.hScrollStatus !== deltaX > 0) {
```
Comment (lijinke666) [inline packages/s2-core/src/facet/base-facet.ts]: 最好统一用 `optimizedDeltaX` 和 `optimizedDeltaY`

https://github.com/antvis/S2/blob/b3d55708bea82142a81dd1d2ede6ff300d5e8957/packages/s2-core/src/facet/base-facet.ts#L1350-L1354

File: packages/s2-core/src/facet/base-facet.ts [hunk 6]
```diff
@@ -1335,6 +1338,28 @@ export abstract class BaseFacet {
   onWheel = (event: WheelEvent) => {
     const { interaction } = this.spreadsheet.options;
     let { deltaX, deltaY, offsetX, offsetY } = event;
+    const { scrollX: currentScrollX, rowHeaderScrollX } =
+      this.getScrollOffset();
+
+    if (this.hScrollStatus !== undefined && this.hScrollStatus !== deltaX > 0) {
+      this.cancelScrollFrame();
+      this.hScrollStatus = deltaX > 0;
```
Comment (lijinke666) [inline packages/s2-core/src/facet/base-facet.ts:1346]: 最好统一用 `optimizedDeltaX` 和 `optimizedDeltaY`

https://github.com/antvis/S2/blob/b3d55708bea82142a81dd1d2ede6ff300d5e8957/packages/s2-core/src/facet/base-facet.ts#L1350-L1354

File: packages/s2-core/src/facet/base-facet.ts [hunk 7]
```diff
@@ -1335,6 +1338,28 @@ export abstract class BaseFacet {
   onWheel = (event: WheelEvent) => {
     const { interaction } = this.spreadsheet.options;
     let { deltaX, deltaY, offsetX, offsetY } = event;
+    const { scrollX: currentScrollX, rowHeaderScrollX } =
+      this.getScrollOffset();
+
+    if (this.hScrollStatus !== undefined && this.hScrollStatus !== deltaX > 0) {
```
Comment (lijinke666) [inline packages/s2-core/src/facet/base-facet.ts:1344]: 应该写在这之后, 不然父容器也会滚动, 需要验证滚动链的问题

https://github.com/antvis/S2/blob/b3d55708bea82142a81dd1d2ede6ff300d5e8957/packages/s2-core/src/facet/base-facet.ts#L1359-L1370

File: packages/s2-core/src/facet/base-facet.ts [hunk 8]
```diff
@@ -180,6 +180,9 @@ export abstract class BaseFacet {
 
   public gridInfo: GridInfo;
 
+  // 标记当前滑动的方向
+  private hScrollStatus: boolean;
```
Comment (lijinke666) [inline packages/s2-core/src/facet/base-facet.ts:184]: 直接用这个吧, 增加一个 `scrollLeft` 和 `scrollRight` 的枚举

https://github.com/antvis/S2/blob/b3d55708bea82142a81dd1d2ede6ff300d5e8957/packages/s2-core/src/facet/base-facet.ts#L226

File: packages/s2-core/src/facet/base-facet.ts [hunk 9]
```diff
@@ -1368,6 +1371,34 @@ export abstract class BaseFacet {
       return;
     }
 
+    if (
+      this.scrollDirection !== undefined &&
+      this.scrollDirection !==
+        (deltaX > 0
```
Comment (lijinke666) [inline packages/s2-core/src/facet/base-facet.ts:1377]: optimizedDeltaX ?

File: packages/s2-core/src/facet/base-facet.ts [hunk 10]
```diff
@@ -1368,6 +1371,34 @@ export abstract class BaseFacet {
...
+        offsetX,
+        offsetY,
+        offset: currentScrollX,
+      });
+
+      return;
+    }
+
+    this.scrollDirection =
+      deltaX > 0 ? ScrollDirection.SCROLL_LEFT : ScrollDirection.SCROLL_RIGHT;
```
Comment (lijinke666) [inline packages/s2-core/src/facet/base-facet.ts:1400]: optimizedDeltaX ?

File: packages/s2-core/src/facet/base-facet.ts (hunk 7 above)
Comment (lijinke666) [reply packages/s2-core/src/facet/base-facet.ts]: 准确来说应该是在 `this.scrollFrameId = requestAnimationFrame` 之前?

image

File: packages/s2-core/src/facet/base-facet.ts [hunk 11]
```diff
@@ -1368,6 +1371,34 @@ export abstract class BaseFacet {
       return;
     }
 
+    if (
```
Comment (lijinke666) [inline packages/s2-core/src/facet/base-facet.ts:1374]: 准确来说应该写在 `this.scrollFrameId = requestAnimationFrame` 之前吧? 看起来 `stopScrollChaining` 也没执行到, 这里也不需要再重复 `this.cancelScrollFrame()` 了

Comment (lijinke666) [reply packages/s2-core/src/facet/base-facet.ts:1374]: ping @DavisYao88

**Comment by lijinke666** [comment]
image

这个问题请看下, 我觉得是有问题的

**Review by lijinke666 — APPROVED** [review]

**Comment by lijinke666** [comment]
下周发布

**Pull request merged**